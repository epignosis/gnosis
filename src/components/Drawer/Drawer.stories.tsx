import React, { useState } from "react";
import { Story } from "@storybook/react";
import { Button } from "../../";
import DrawerComponent, { DrawerProps } from "./Drawer";

export default {
  title: "components/Drawer",
  component: DrawerComponent,
  argTypes: {
    placement: {
      control: {
        type: "select",
        options: ["left", "right"],
      },
    },
  },
  decorators: [
    (Story: Story): JSX.Element => (
      <div
        style={{
          width: "100%",
          height: "100%",
          display: "flex",
          justifyContent: "center",
          marginTop: 200,
        }}
      >
        <DrawerComponent.Root />
        <Story />
      </div>
    ),
  ],
};

type DrawerArgs = Pick<DrawerProps, "placement" | "showMask"> & {
  headerCloseBtn: boolean;
};

export const Default: Story<DrawerArgs> = (args) => {
  const { headerCloseBtn, ...rest } = args;
  const [isOpen, setIsOpen] = useState(false);

  return (
    <>
      <Button onClick={(): void => setIsOpen(true)}>Open Drawer</Button>
      <DrawerComponent isOpen={isOpen} onClose={(): void => setIsOpen(false)} {...rest}>
        <DrawerComponent.Header closable={headerCloseBtn}>Drawer title</DrawerComponent.Header>
        <DrawerComponent.Body>
          <div style={{ padding: "2rem" }}>
            <p>Drawer content</p>
            <ul>
              <li>List item 1</li>
              <li>List item 2</li>
              <li>List item 3</li>
            </ul>
          </div>
        </DrawerComponent.Body>
        <DrawerComponent.Footer>
          <div style={{ display: "flex", justifyContent: "center" }}>
            <Button onClick={(): void => setIsOpen(false)}>Close Drawer</Button>
          </div>
        </DrawerComponent.Footer>
      </DrawerComponent>
    </>
  );
};

Default.args = {
  headerCloseBtn: true,
  placement: "left",
  showMask: true,
};
