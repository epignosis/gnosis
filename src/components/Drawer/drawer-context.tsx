import React, { createContext, useContext, useReducer, ReactNode } from "react";

type State = { showMask: boolean; drawers: { id: string; isOpen: boolean }[] };
type Action = { type: "open"; id: string } | { type: "close"; id: string };
type Dispatch = (action: Action) => void;
type DrawerProviderProps = { children: ReactNode };
type DrawerContext = { state: State; dispatch: Dispatch };

const DrawersStateContext = createContext<DrawerContext | null>(null);

const INITIAL_STATE = {
  drawers: [],
  showMask: false,
};

const drawersReducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "open":
      return {
        showMask: true,
        drawers: [
          ...state.drawers,
          {
            id: action.id,
            isOpen: true,
          },
        ],
      };
    case "close": {
      const drawers = state.drawers.filter(({ id }) => id !== action.id);
      const showMask = drawers.reduce((acc, val) => {
        if (acc) return true;
        if (val.isOpen) return true;
        return false;
      }, false);

      return {
        drawers,
        showMask,
      };
    }
    default:
      return INITIAL_STATE;
  }
};

const DrawerProvider = ({ children }: DrawerProviderProps): JSX.Element => {
  const [state, dispatch] = useReducer(drawersReducer, INITIAL_STATE);

  return (
    <DrawersStateContext.Provider value={{ state, dispatch }}>
      {children}
    </DrawersStateContext.Provider>
  );
};

const useDrawersContext = (): DrawerContext => {
  const context = useContext(DrawersStateContext);

  if (!context) throw new Error("`useDrawersContext` must be used within a DrawerProvider");

  return context;
};

export { DrawerProvider, useDrawersContext };
