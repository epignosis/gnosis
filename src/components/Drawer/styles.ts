import { css, SerializedStyles } from "@emotion/react";

export const drawerContainer = (width: string): SerializedStyles => {
  return css`
    position: fixed;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
    z-index: 1000;

    .dialog {
      position: absolute;
      height: 100%;
      width: 100%;
      max-width: ${width};
      display: flex;
      flex-direction: column;
      background-color: #ffffff;
      padding: 0;
      border: 0;

      &.placement-left {
        left: 0;
        right: auto;
      }

      &.placement-right {
        left: auto;
        right: 0;
      }
    }
  `;
};
