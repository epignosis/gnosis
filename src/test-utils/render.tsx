import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
