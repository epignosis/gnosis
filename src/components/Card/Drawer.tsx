import React, { FC, ReactNode } from "react";
import { m, AnimatePresence, Variants, Transition } from "framer-motion";
import { useClickAway } from "@umijs/hooks";
import { CloseSVG } from "../../icons/";
import { drawerContainer } from "./styles";

export type DrawerProps = FC<{
  isOpen: boolean;
  onClose: () => void;
  title?: ReactNode;
  footer?: ReactNode;
}>;

const transition: Transition = {
  duration: 0.25,
  ease: "easeInOut",
};

const drawerContainerVariants: Variants = {
  open: {
    width: "100%",
    transition: {
      ...transition,
      when: "beforeChildren",
    },
    transitionEnd: {
      overflowY: "auto",
    },
  },
  closed: {
    width: "0%",
    overflowY: "hidden",
    transition: {
      ...transition,
      when: "afterChildren",
    },
  },
};

const drawerContentVariants: Variants = {
  open: {
    opacity: 1,
  },
  closed: {
    opacity: 0,
  },
};

const Drawer: DrawerProps = ({ isOpen, onClose, title, footer, children }) => {
  const containerRef = useClickAway((): void => {
    if (isOpen) {
      onClose();
    }
  });

  return (
    <AnimatePresence>
      {isOpen && (
        <m.section
          css={drawerContainer}
          ref={containerRef}
          initial="closed"
          animate={isOpen && "open"}
          exit="closed"
          variants={drawerContainerVariants}
        >
          <m.div className="content-wrapper" variants={drawerContentVariants}>
            <div>
              <header>
                <div>{title}</div>
                <button onClick={onClose}>
                  <CloseSVG height={16} />
                </button>
              </header>
              <p>{children}</p>
            </div>
            {footer && <footer>{footer}</footer>}
          </m.div>
        </m.section>
      )}
    </AnimatePresence>
  );
};

export default Drawer;
