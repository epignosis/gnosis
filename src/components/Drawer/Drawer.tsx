import React, { Children, cloneElement, FC, ReactElement, useEffect } from "react";
import { createPortal } from "react-dom";
import { AnimatePresence, m, Variants, MotionStyle, domAnimation, LazyMotion } from "framer-motion";
import classNames from "classnames";
import { drawerContainer } from "./styles";
import Header, { HeaderProps } from "./components/Header";
import Body from "./components/Body";
import Mask from "./components/Mask";
import Footer from "./components/Footer";
// import { useTrap } from "@hooks";

type DialogVariants = {
  placement: "left" | "right";
  width: string;
};

const dialogVariants: Variants = {
  initial: ({ placement, width }: DialogVariants) => ({
    opacity: 0,
    x: placement === "right" ? "100%" : `-${width}`,
    transition: {
      type: "tween",
    },
  }),
  expanded: () => ({
    opacity: 1,
    x: 0,
    transition: {
      type: "tween",
    },
  }),
};

const DRAWER_ROOT = "drawerRoot";

const DrawerRoot: FC = () => <div id={DRAWER_ROOT} />;

export type DrawerProps = {
  isOpen: boolean;
  onClose: () => void;
  showMask?: boolean;
  placement?: "left" | "right";
  width?: string;
  dialogStyles?: MotionStyle;
};

type DrawerCompoundProps = {
  Header: FC<Omit<HeaderProps, "onClose"> & { closable?: boolean }>;
  Body: FC;
  Footer: FC;
  Root: FC;
};

const Drawer: FC<DrawerProps> & DrawerCompoundProps = (props) => {
  const {
    isOpen,
    onClose,
    placement = "left",
    showMask = true,
    width = "31.5rem",
    dialogStyles,
    children,
  } = props;
  // const dialogElement = useTrap<HTMLDivElement>(isOpen);
  const clonedChildren = Children.map(children, (child) => {
    return (
      child &&
      cloneElement(child as ReactElement, {
        onClose: (child as ReactElement).props?.closable && onClose,
      })
    );
  });
  const drawerEl = document.getElementById(DRAWER_ROOT);
  const dialogClassNames = classNames({
    dialog: true,
    "placement-left": placement === "left",
    "placement-right": placement === "right",
  });
  const drawer = (
    <LazyMotion features={domAnimation}>
      <AnimatePresence>
        {isOpen && (
          <div css={drawerContainer(width)}>
            {showMask && <Mask onClose={onClose} />}
            <m.div
              id="drawer-dialog"
              style={dialogStyles}
              className={dialogClassNames}
              // ref={dialogElement}
              aria-expanded={isOpen}
              aria-hidden={!isOpen}
              aria-modal="true"
              initial="initial"
              animate="expanded"
              exit="initial"
              variants={dialogVariants}
              custom={{ placement, width }}
            >
              {clonedChildren}
            </m.div>
          </div>
        )}
      </AnimatePresence>
    </LazyMotion>
  );

  useEffect(() => {
    if (isOpen) {
      (document.querySelector("body") as HTMLBodyElement).style.overflow = "hidden";
    } else {
      (document.querySelector("body") as HTMLBodyElement).style.overflow = "";
    }
  }, [isOpen]);

  return drawerEl && createPortal(drawer, drawerEl);
};

Drawer.Header = Header;
Drawer.Body = Body;
Drawer.Footer = Footer;
Drawer.Root = DrawerRoot;

export default Drawer;
