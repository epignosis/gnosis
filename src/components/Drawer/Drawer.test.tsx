import React from "react";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import Drawer from "./Drawer";
import { screen, render, cleanup } from "@test-utils/render";

describe("<Drawer/>", () => {
  beforeEach(() => {
    const drawerEl = document.createElement("div");
    drawerEl.setAttribute("id", "drawerRoot");
    document.body.appendChild(drawerEl);
  });

  afterEach(() => {
    cleanup();
  });

  it("renders correctly", () => {
    const headerTxt = faker.unique(faker.lorem.sentence);
    const bodyTxt = faker.unique(faker.lorem.sentence);
    const footerTxt = faker.unique(faker.lorem.sentence);
    const mockFn = jest.fn();

    render(
      <Drawer isOpen onClose={mockFn}>
        <Drawer.Header>{headerTxt}</Drawer.Header>
        <Drawer.Body>{bodyTxt}</Drawer.Body>
        <Drawer.Footer>{footerTxt}</Drawer.Footer>
      </Drawer>,
    );

    const header = screen.getByText(headerTxt);
    const body = screen.getByText(bodyTxt);
    const footer = screen.getByText(footerTxt);
    const mask = screen.getByTestId("mask");

    userEvent.click(mask);

    expect(header).toHaveTextContent(headerTxt);
    expect(body).toHaveTextContent(bodyTxt);
    expect(footer).toHaveTextContent(footerTxt);
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it("renders correctly without Header and Footer", () => {
    const bodyTxt = faker.lorem.word();
    const mockFn = jest.fn();

    render(
      <Drawer isOpen onClose={mockFn}>
        <Drawer.Body>{bodyTxt}</Drawer.Body>
      </Drawer>,
    );

    const header = screen.queryByTestId("drawer-header");
    const body = screen.getByText(bodyTxt);
    const footer = screen.queryByTestId("drawer-footer");

    expect(header).not.toBeInTheDocument();
    expect(body).toHaveTextContent(bodyTxt);
    expect(footer).not.toBeInTheDocument();
  });

  it("renders correctly with close button", () => {
    const headerTxt = faker.lorem.word();
    const mockFn = jest.fn();

    const { rerender } = render(
      <Drawer isOpen onClose={mockFn}>
        <Drawer.Header closable>{headerTxt}</Drawer.Header>
      </Drawer>,
    );

    const header = screen.getByText(headerTxt);
    const closeBtn = screen.getByRole("button");

    userEvent.click(closeBtn);

    expect(header).toHaveTextContent(headerTxt);
    expect(closeBtn).toBeInTheDocument();
    expect(mockFn).toHaveBeenCalledTimes(1);

    rerender(
      <Drawer isOpen={false} onClose={mockFn}>
        <Drawer.Header closable>{headerTxt}</Drawer.Header>
      </Drawer>,
    );

    expect(header).not.toBeVisible();
  });

  it("Header renders with JSX content", () => {
    const headerTxt = faker.lorem.word();
    const mockFn = jest.fn();

    render(
      <Drawer isOpen onClose={mockFn}>
        <Drawer.Header closable>
          <button>{headerTxt}</button>
        </Drawer.Header>
      </Drawer>,
    );

    const header = screen.getByText(headerTxt);

    expect(header).toBeInTheDocument();
    expect(header).toHaveTextContent(headerTxt);
  });

  it("matches snapshot with header, body and footer", () => {
    render(
      <Drawer isOpen onClose={jest.fn()}>
        <Drawer.Header>Test header</Drawer.Header>
        <Drawer.Body>Test body</Drawer.Body>
        <Drawer.Footer>Test footer</Drawer.Footer>
      </Drawer>,
    );
    const drawer = screen.getByTestId("drawer");

    expect(drawer).toMatchSnapshot();
  });

  it("matches snapshot without header and footer", () => {
    render(
      <Drawer isOpen onClose={jest.fn()}>
        <Drawer.Body>Test body</Drawer.Body>
      </Drawer>,
    );
    const drawer = screen.getByTestId("drawer");

    expect(drawer).toMatchSnapshot();
  });

  it("matches snapshot with close button", () => {
    render(
      <Drawer isOpen onClose={jest.fn()}>
        <Drawer.Header closable>Title</Drawer.Header>
      </Drawer>,
    );
    const drawer = screen.getByTestId("drawer");

    expect(drawer).toMatchSnapshot();
  });

  it("matches snapshot with noGutters Header", () => {
    render(
      <Drawer isOpen onClose={jest.fn()}>
        <Drawer.Header noGutters>Title</Drawer.Header>
      </Drawer>,
    );
    const drawer = screen.getByTestId("drawer");

    expect(drawer).toMatchSnapshot();
  });

  it("matches snapshot with right placement", () => {
    render(
      <Drawer isOpen onClose={jest.fn()} placement="right">
        <Drawer.Header>Title</Drawer.Header>
      </Drawer>,
    );
    const drawer = screen.getByTestId("drawer");

    expect(drawer).toMatchSnapshot();
  });
});
