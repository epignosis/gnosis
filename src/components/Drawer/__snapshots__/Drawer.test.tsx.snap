// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer/> matches snapshot with close button 1`] = `
<div
  class="css-1jowdg0-drawerContainer-drawerContainer"
  data-testid="drawer"
>
  <div
    class="css-tigpv4-maskContainer"
    data-testid="mask"
    style="opacity: 0;"
  />
  <div
    aria-expanded="true"
    aria-hidden="false"
    aria-modal="true"
    class="dialog placement-left"
    id="drawer-dialog"
    style="opacity: 0; transform: none;"
  >
    <header
      class="css-nabu8w-drawerHeader-drawerHeader"
      data-testid="drawer-header"
      id="drawer-title"
    >
      <h2
        class="css-1r6ji88-headingContainer-headingContainer"
      >
        Title
      </h2>
      <button
        aria-label="Close drawer"
        type="button"
      >
        <svg-file-stub
          height="24"
        />
      </button>
    </header>
  </div>
</div>
`;

exports[`<Drawer/> matches snapshot with header, body and footer 1`] = `
<div
  class="drawer css-1jowdg0-drawerContainer-drawerContainer"
  data-testid="drawer"
  id="main-drawer"
>
  <div
    class="css-tigpv4-maskContainer"
    data-testid="mask"
    style="opacity: 0;"
  />
  <div
    aria-expanded="true"
    aria-hidden="false"
    aria-modal="true"
    class="dialog placement-left"
    id="drawer-dialog"
    style="opacity: 0; transform: none;"
  >
    <header
      class="css-nabu8w-drawerHeader-drawerHeader"
      data-testid="drawer-header"
      id="drawer-title"
    >
      <h2
        class="css-1r6ji88-headingContainer-headingContainer"
      >
        Test header
      </h2>
    </header>
    <div
      class="css-1vbxgux-drawerBody"
    >
      Test body
    </div>
    <footer
      class="css-1ppo1yc-footerContainer-footerContainer"
      data-testid="drawer-footer"
    >
      Test footer
    </footer>
  </div>
</div>
`;

exports[`<Drawer/> matches snapshot with noGutters Header 1`] = `
<div
  class="css-1jowdg0-drawerContainer-drawerContainer"
  data-testid="drawer"
>
  <div
    class="css-tigpv4-maskContainer"
    data-testid="mask"
    style="opacity: 0;"
  />
  <div
    aria-expanded="true"
    aria-hidden="false"
    aria-modal="true"
    class="dialog placement-left"
    id="drawer-dialog"
    style="opacity: 0; transform: none;"
  >
    <header
      class="css-1uz1e9a-drawerHeader-drawerHeader"
      data-testid="drawer-header"
      id="drawer-title"
    >
      <h2
        class="css-1r6ji88-headingContainer-headingContainer"
      >
        Title
      </h2>
    </header>
  </div>
</div>
`;

exports[`<Drawer/> matches snapshot with right placement 1`] = `
<div
  class="css-1jowdg0-drawerContainer-drawerContainer"
  data-testid="drawer"
>
  <div
    class="css-tigpv4-maskContainer"
    data-testid="mask"
    style="opacity: 0;"
  />
  <div
    aria-expanded="true"
    aria-hidden="false"
    aria-modal="true"
    class="dialog placement-right"
    id="drawer-dialog"
    style="opacity: 0; transform: translateX(100%) translateZ(0);"
  >
    <header
      class="css-nabu8w-drawerHeader-drawerHeader"
      data-testid="drawer-header"
      id="drawer-title"
    >
      <h2
        class="css-1r6ji88-headingContainer-headingContainer"
      >
        Title
      </h2>
    </header>
  </div>
</div>
`;

exports[`<Drawer/> matches snapshot without header and footer 1`] = `
<div
  class="css-1jowdg0-drawerContainer-drawerContainer"
  data-testid="drawer"
>
  <div
    class="css-tigpv4-maskContainer"
    data-testid="mask"
    style="opacity: 0;"
  />
  <div
    aria-expanded="true"
    aria-hidden="false"
    aria-modal="true"
    class="dialog placement-left"
    id="drawer-dialog"
    style="opacity: 0; transform: none;"
  >
    <div
      class="css-1vbxgux-drawerBody"
    >
      Test body
    </div>
  </div>
</div>
`;
